File: src/ast/ast.go
package ast

import (
	"bytes"
	"strconv"
	"strings"
)

// Two Distinct enum like types
type MathState uint8
type ArgNeed uint8

const (
	TextState MathState = iota
	InlineState
)

const (
	MainArg ArgNeed = iota
	Optional
	StarArg
)

// Node and Statement interface signatures
type Node interface {
	Type() string
	String() string
}

type Statement interface {
	Node
	statementNode()
}

////////////////////////////////////////

// Latex Statement
type Latex struct {
	Stmts []Statement
}

func (latex *Latex) String() string {
	var out bytes.Buffer

	for _, stmt := range latex.Stmts {
		out.WriteString(stmt.String())
	}

	return out.String()
}

// DocumentClass Statement
type DocumentClass struct {
	Name    string
	Options []Latex
}

func (doc *DocumentClass) String() string {
	var out bytes.Buffer

	out.WriteString("\\documentclass")
	if len(doc.Options) != 0 {
		var opts []string
		out.WriteByte('[')
		for _, opt := range doc.Options {
			opts = append(opts, opt.String())
		}
		out.WriteString(strings.Join(opts, ","))
		out.WriteByte(']')
	}

	out.WriteByte('{')
	out.WriteString(doc.Name)
	out.WriteString("}\n")

	return out.String()
}

// Usepackage Statement
type UsePackage struct {
	Name    string
	Options []Latex
}

func (up *UsePackage) String() string {
	var out bytes.Buffer

	out.WriteString("\\usepackage")
	if len(up.Options) != 0 {
		var opts []string
		out.WriteByte('[')
		for _, opt := range up.Options {
			opts = append(opts, opt.String())
		}
		out.WriteString(strings.Join(opts, ","))
		out.WriteByte(']')
	}

	out.WriteByte('{')
	out.WriteString(up.Name)
	out.WriteString("}\n")

	return out.String()
}

// MultiUsePackages Statement
type MultiUsePackages struct {
	Pkgs []Statement
}

func (mup *MultiUsePackages) String() string {
	var out bytes.Buffer

	for _, pkg := range mup.Pkgs {
		out.WriteString(pkg.String())
	}

	return out.String()
}

// DocumentStart Statement
type DocumentStart struct{}

func (ds *DocumentStart) String() string { return "\n\\begin{document}\n" }

// DocumentEnd Statement
type DocumentEnd struct{}

func (de *DocumentEnd) String() string { return "\n\\end{document}\n" }

// MainText Statement
type MainText struct {
	Value string
}

func (mt *MainText) String() string { return mt.Value }

// Integer Statement
type Integer struct {
	Value int64
}

func (i *Integer) String() string { return strconv.FormatInt(i.Value, 10) }

// Float Statement
type Float struct {
	Value float64
}

func (f *Float) String() string { return strconv.FormatFloat(f.Value, 'g', -1, 64) }

type RawLatex struct {
	Value string
}

func (rl *RawLatex) String() string { return rl.Value }

type MathText struct {
	State MathState
	Text  []Statement
}

func (mt *MathText) String() string {
	var out bytes.Buffer

	switch mt.State {
	case TextState:
		out.WriteByte('$')
		for _, text := range mt.Text {
			out.WriteString(text.String())
		}
		out.WriteByte('$')
	case InlineState:
		out.WriteString("\\[")
		for _, text := range mt.Text {
			out.WriteString(text.String())
		}
		out.WriteString("\\]")
	}

	return out.String()
}

type PlainTextInMath struct {
	Value *Latex
}

func (ptlm *PlainTextInMath) String() string {
	var out bytes.Buffer

	out.WriteString("\\text{")
	out.WriteString(ptlm.Value.String())
	out.WriteByte('\b')
	out.WriteByte('}')

	return out.String()
}

// LatexFunction Statement
type Argument struct {
	ArgType ArgNeed
	Text    *Latex
}
type LatexFunction struct {
	Name string
	Args []Argument
}

func (lf *LatexFunction) String() string {
	var out bytes.Buffer

	out.WriteByte('\\')
	out.WriteString(lf.Name)

	for _, arg := range lf.Args {
		switch arg.ArgType {
		case MainArg:
			out.WriteByte('{')
			out.WriteString(arg.Text.String())
			out.WriteByte('}')
		case Optional:
			out.WriteByte('[')
			out.WriteString(arg.Text.String())
			out.WriteByte(']')
		case StarArg:
			out.WriteByte('*')
		}
	}

	return out.String()
}

// Environment Statement
type Environment struct {
	Name string
	Args []Argument
	Text *Latex
}

func (env *Environment) String() string {
	var out bytes.Buffer

	out.WriteString("\\begin{")
	out.WriteString(env.Name)
	out.WriteByte('}')

	for _, arg := range env.Args {
		switch arg.ArgType {
		case MainArg:
			out.WriteByte('{')
			out.WriteString(arg.Text.String())
			out.WriteByte('}')
		case Optional:
			out.WriteByte('[')
			out.WriteString(arg.Text.String())
			out.WriteByte(']')
		case StarArg:
			out.WriteByte('*')
		}
	}

	out.WriteString(env.Text.String())

	out.WriteString("\\end{")
	out.WriteString(env.Name)
	out.WriteString("}\n")

	return out.String()
}

//////////////////////////////////////////////////
// Implementation Dummy
//////////////////////////////////////////////////

func (latex *Latex) Type() string          { return "LaTeX" }
func (doc *DocumentClass) Type() string    { return "DocumentClass" }
func (up *UsePackage) Type() string        { return "UsePackage" }
func (mup *MultiUsePackages) Type() string { return "MultiUsePackages" }
func (ds *DocumentStart) Type() string     { return "DocumentStart" }
func (de *DocumentEnd) Type() string       { return "DocumentEnd" }
func (mt *MainText) Type() string          { return "MainText" }
func (i *Integer) Type() string            { return "Integer" }
func (f *Float) Type() string              { return "Float" }
func (rl *RawLatex) Type() string          { return "RawLatex" }
func (mt *MathText) Type() string          { return "MathText" }
func (ptlm *PlainTextInMath) Type() string { return "PlainTextInMath" }
func (lf *LatexFunction) Type() string     { return "LatexFunction" }
func (env *Environment) Type() string      { return "Environment" }

func (doc *DocumentClass) statementNode()    {}
func (up *UsePackage) statementNode()        {}
func (mup *MultiUsePackages) statementNode() {}
func (ds *DocumentStart) statementNode()     {}
func (de *DocumentEnd) statementNode()       {}
func (mt *MainText) statementNode()          {}
func (i *Integer) statementNode()            {}
func (f *Float) statementNode()              {}
func (rl *RawLatex) statementNode()          {}
func (mt *MathText) statementNode()          {}
func (ptlm *PlainTextInMath) statementNode() {}
func (lf *LatexFunction) statementNode()     {}
func (env *Environment) statementNode()      {}

File: src/ast.zig
const std = @import("std");
const token = @import("token.zig");

const ArrayList = std.ArrayList;
const Allocator = std.mem.Allocator;

pub const GenerateAstError = error{
    AppendToLatexFailed,
    CloneLatexFailed,
};

pub const Latex = struct {
    stmts: ArrayList(Statement),

    const Self = @This();

    pub fn init(allocator: Allocator) Self {
        return .{ .stmts = ArrayList(Statement).init(allocator) };
    }

    pub fn deinit(self: Self) void {
        for (self.stmts.items) |*stmt| {
            stmt.deinit();
        }
        self.stmts.deinit();
    }

    pub fn append(self: *Self, stmt: Statement) GenerateAstError!void {
        self.stmts.append(stmt) catch return error.AppendToLatexFailed;
    }

    pub fn len(self: *Self) usize {
        return self.stmts.len;
    }
};

pub const StatementType = enum(u8) {
    document_class,
    use_packages,
    document_start,
    document_end,
    main_text,
    integer,
    float,
    raw_latex,
    math_text,
    latex_function,
    plain_text_in_math,
    environment,
    phantom_begin_environment,
    phantom_end_environment,
    function_define,
    environment_define,
};

pub const Statement = union(StatementType) {
    document_class: Package,
    use_packages: ArrayList(Package),
    document_start: void,
    document_end: void,
    main_text: []const u8,
    integer: i64,
    float: f64,
    raw_latex: []const u8,
    math_text: MathText,
    plain_text_in_math: PlainTextInMath,
    latex_function: LatexFnt,
    environment: Environment,
    phantom_begin_environment: PhantomBeginEnv,
    phantom_end_environment: ArrayList(u8),
    function_define: FunctionDefine,
    environment_define: EnvironmentDefine,

    pub fn deinit(self: @This()) void {
        switch (self) {
            .document_class => |*class| class.deinit(),
            .use_packages => |*pkgs| {
                for (pkgs.items) |*pkg| {
                    pkg.deinit();
                }
                pkgs.deinit();
            },
            .math_text => |*math_text| math_text.deinit(),
            .plain_text_in_math => |*plain_text| plain_text.deinit(),
            .latex_function => |*fnt| fnt.deinit(),
            .environment => |*env| env.deinit(),
            .phantom_begin_environment => |*env| env.deinit(),
            .phantom_end_environment => |*env| env.deinit(),
            .function_define => |*fnt_def| fnt_def.deinit(),
            .environment_define => |*env_def| env_def.deinit(),
            else => {},
        }
    }
};

pub const Package = struct {
    name: ArrayList(u8),
    options: ?ArrayList(Latex),

    pub fn init(alloc: Allocator, comptime has_option: bool) @This() {
        if (has_option) {
            return .{
                .name = undefined,
                .options = ArrayList(Latex).init(alloc),
            };
        } else {
            return .{
                .name = undefined,
                .options = null,
            };
        }
    }

    pub fn deinit(self: @This()) void {
        self.name.deinit();

        if (self.options) |options| {
            for (options.items) |*latex| {
                latex.deinit();
            }
            options.deinit();
        }
    }
};

pub const MathText = struct {
    state: MathState,
    text: Latex,

    pub fn init(alloc: Allocator) @This() {
        return .{ .state = undefined, .text = Latex.init(alloc) };
    }

    pub fn deinit(self: @This()) void {
        self.text.deinit();
    }
};

pub const PlainTextInMath = struct {
    trim: TrimWhitespace,
    text: Latex,

    pub fn deinit(self: @This()) void {
        self.text.deinit();
    }
};

pub const LatexFnt = struct {
    name: []const u8,
    args: ArrayList(Argument),
    has_space: bool,

    pub fn deinit(self: @This()) void {
        for (self.args.items) |*arg| {
            arg.deinit();
        }
        self.args.deinit();
    }
};

pub const Environment = struct {
    name: ArrayList(u8),
    args: ArrayList(Argument),
    text: Latex,

    pub fn deinit(self: @This()) void {
        self.name.deinit();
        for (self.args.items) |arg| {
            arg.deinit();
        }
        self.args.deinit();
        self.text.deinit();
    }
};

pub const PhantomBeginEnv = struct {
    name: ArrayList(u8),
    args: ArrayList(Argument),

    pub fn deinit(self: @This()) void {
        self.name.deinit();
        for (self.args.items) |arg| {
            arg.deinit();
        }
        self.args.deinit();
    }
};

pub const FunctionDefine = struct {
    style: FunctionStyle,
    name: ArrayList(u8),
    args: ArrayList(u8),
    trim: TrimWhitespace,
    body: Latex,

    pub fn deinit(self: @This()) void {
        self.name.deinit();
        self.args.deinit();
        self.body.deinit();
    }
};

pub const EnvironmentDefine = struct {
    is_redefine: bool,
    name: ArrayList(u8),
    args_num: u8,
    optional_arg: ?Latex,
    trim: TrimWhitespace,
    begin_part: Latex,
    end_part: Latex,

    pub fn deinit(self: @This()) void {
        self.name.deinit();
        self.begin_part.deinit();
        self.end_part.deinit();
        if (self.optional_arg) |optional| {
            optional.deinit();
        }
    }
};

pub const Argument = struct {
    arg_type: ArgNeed,
    inner: Latex,

    pub fn init(alloc: Allocator) @This() {
        return .{ .arg_type = undefined, .inner = Latex.init(alloc) };
    }

    pub fn deinit(self: @This()) void {
        self.inner.deinit();
    }
};

pub const TrimWhitespace = struct {
    start: bool,
    end: bool,
    mid: ?bool,
};

pub const MathState = enum(u2) {
    text_math,
    display_math,
};

pub const ArgNeed = enum(u2) {
    main_arg,
    optional,
    star_arg,
};

pub const FunctionStyle = enum(u8) {
    plain,
    long_plain,
    outer_plain,
    long_outer_plain,
    expand,
    long_expand,
    outer_expand,
    long_outer_expand,
    global,
    long_global,
    outer_global,
    long_outer_global,
    expand_global,
    long_expand_global,
    outer_expand_global,
    long_outer_expand_global,

    pub inline fn default() @This() {
        return .plain;
    }

    pub fn from_toktype(toktype: token.Type) ?@This() {
        return switch (toktype) {
            .function_def => .plain,
            .long_function_def => .long_plain,
            .outer_function_def => .outer_plain,
            .long_outer_function_def => .long_outer_plain,
            .e_function_def => .expand,
            .e_long_function_def => .long_expand,
            .e_outer_function_def => .outer_expand,
            .e_long_outer_function_def => .long_outer_expand,
            .g_function_def => .global,
            .g_long_function_def => .long_global,
            .g_outer_function_def => .outer_global,
            .g_long_outer_function_def => .long_outer_global,
            .x_function_def => .expand_global,
            .x_long_function_def => .long_expand_global,
            .x_outer_function_def => .outer_expand_global,
            .x_long_outer_function_def => .long_outer_expand_global,
            else => null,
        };
    }
};
